exoscale/api/compute.py

```diff
@@ -0,0 +1,1714 @@
+        description (str): the Security Group rule description
+        network_cidr (str): a source/destination network CIDR to match
+            (conflicts with security_group)
+        security_group (SecurityGroup): a source/destination Security Group to match
+            (conflicts with network_cidr)
+        port (str): the source/destination port or port range to match
+        protocol (str): a network protocol to match
+        icmp_code (int): an ICMP code to match
+        icmp_type (int): an ICMP type to match
+    """
+
+    type = attr.ib()
+    id = attr.ib(default=None)
+    description = attr.ib(default=None, repr=False)
+    network_cidr = attr.ib(default=None, repr=False)
+    security_group = attr.ib(default=None, repr=False)
+    port = attr.ib(default=None, repr=False)
+    protocol = attr.ib(default="tcp", repr=False)
+    icmp_code = attr.ib(default=None, repr=False)
+    icmp_type = attr.ib(default=None, repr=False)
+
+    @classmethod
+    def ingress(cls, **kwargs):
+        return SecurityGroupRule(type="ingress", **kwargs)
+
+    @classmethod
+    def egress(cls, **kwargs):
+        return SecurityGroupRule(type="egress", **kwargs)
+
+    def _parse_port(self):
+        """
+        Parse the rule port attribute and returns a port range.
+
+        Returns:
+            int: start port
+            int: end port
+        """
+
+        if not self.port:
+            return None, None
+
+        start_port, end_port = (
+            self.port.split("-", maxsplit=1) if "-" in self.port else (self.port, None)
+        )
+        if not end_port:
+            end_port = start_port
+
+        return start_port, end_port
+
+
+@attr.s
+class SSHKey(Resource):
+    """
+    A SSH key.
+
+    Attributes:
+        name (str): the SSH key unique name
+        fingerprint (str): the SSH key fingerprint
+        private_key (str): the SSH private key, or None if registered SSH key
+    """
+
+    res = attr.ib(repr=False)
+    name = attr.ib()
+    fingerprint = attr.ib()
+    private_key = attr.ib(default=None, repr=False)
+
+    @classmethod
+    def from_cs(cls, res):
+        return SSHKey(
+            res,
+            name=res["name"],
+            fingerprint=res["fingerprint"],
+            private_key=res.get("privatekey", None),
+        )
+
+
+@attr.s
+class Zone(Resource):
+    """
+    An Exoscale zone.
+
+    Attributes:
+        id (str): the zone unique identifier
+        name (str): the zone name
+    """
+
+    res = attr.ib(repr=False)
+    id = attr.ib()
+    name = attr.ib()
+
+    @classmethod
+    def from_cs(cls, res):
+        return Zone(res, id=res["id"], name=res["name"])
+
+
+class ComputeAPI(API):
+    """
+    An Exoscale Compute API call.
+
+    Attributes:
+        endpoint (str): the Compute API endpoint
+        key (str): the Compute API key
+        secret (str): the Compute API secret
+    """
+
+    def __init__(self, endpoint="", key="", secret=""):
```

Comment:
Changed in e9bc32f.

Reply:
Actually I have to revert this change because it doesn't work: since the caller actually provides a value (even if it's `None`) the default arg value is never picked up, so we end up passing `cs.CloudStack(endpoint=None)`.